cpp/cmake_modules/ThirdpartyToolchain.cmake

```diff
@@ -671,6 +671,18 @@ if (ARROW_WITH_ZLIB)
 # ZLIB
 
   if("${ZLIB_HOME}" STREQUAL "")
+    if(ARROW_ZLIB_VENDORED)
+      set(ZLIB_FOUND FALSE)
+    else()
+      find_package(ZLIB REQUIRED)
+    endif()
+  else()
+    find_package(ZLIB REQUIRED)
+  endif()
+  if(ZLIB_FOUND)
+    set(ZLIB_VENDORED 0)
+  else()
+    set(ZLIB_VENDORED 1)
```

Comment:
Similarly, we should make it an error if ARROW_ZLIB_VENDORED is false and a third-party zlib couldn't be found.

Reply:
Actually, this sounds reasonable since building needs the development headers, which aren't always installed (as opposed to the library itself, which is pretty much everywhere).